torch/csrc/jit/script/compiler.cpp

```diff
@@ -304,6 +304,27 @@ std::shared_ptr<SugaredValue> emitBuiltinCall(
   for(size_t i = 0; i < op->num_outputs; ++i)
     n->addOutput();
 
+  // special handling for the tuple that cat takes as its first argument
+  if(name == "cat") {
+    ensureTensors(loc, inputs.slice(1));
+    auto first = inputs.at(0);
+    if(first->type()->kind() != TupleType::Kind) {
+      throw ErrorReport(loc) << "expected a tuple";
+    }
+    if(inputs.size() + attributes.size() > 2) {
+      throw ErrorReport(loc) << "expected at most 2 inputs";
+    }
+    // flatten the tuple into the argument list
+    auto unpacked = graph->insertNode(graph->createTupleUnpack(first));
+    ensureTensors(loc, unpacked->outputs());
+    n->removeInput(0);
+    for(size_t i = 0; i < unpacked->outputs().size(); ++i) {
+      n->insertInput(i, unpacked->outputs().at(i));
+    }
+  } else {
+    ensureTensors(loc, inputs);
```

Comment:
Uh `ensureTensors` is a bad name, because it also accepts `Dynamic`. This is what we want here, but it will be very easy to get it wrong later.

Reply:
I have a future PR that is going to change this into a subtyping relationship. I'll fix this then.